tools/driver/p4c/main.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+
+"""
+p4c - P4 Compiler Driver
+
+"""
+
+from __future__ import absolute_import
+import argparse
+import glob
+import itertools
+import os
+import subprocess
+import sys
+
+import p4c.run
+import p4c.util
+import p4c.config
+
+commands = {}
+
+def init_globals():
+    commands['preprocessor'] = []
+    commands['compiler'] = []
+    commands['assembler'] = []
+    commands['linker'] = []
+
+def save_options(cmd, options):
+    for opt in options:
+        opt_split = opt.split(' ')
+        for s in opt_split:
+            commands[cmd].append(s)
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-v", "--version", dest="show_version",
+                        help="show version and exit",
+                        action="store_true", default=False)
+    parser.add_argument("-debug", dest="debug",
+                        help="verbose",
+                        action="store_true", default=False)
+    parser.add_argument("-###", dest="dry_run",
+                        help="print (but do not run) the commands",
+                        action="store_true", default=False)
+    parser.add_argument("-Xpreprocessor", dest="preprocessor_options",
+                        metavar="<arg>",
+                        help="Pass <arg> to the preprocessor",
+                        action="append", default=[])
+    parser.add_argument("-Xp4c", dest="compiler_options",
+                        metavar="<arg>",
+                        help="Pass <arg> to the compiler",
+                        action="append", default=[])
+    parser.add_argument("-Xassembler", dest="assembler_options",
+                        metavar="<arg>",
+                        help="Pass <arg> to the assembler",
+                        action="append", default=[])
+    parser.add_argument("-Xlinker", dest="linker_options",
+                        metavar="<arg>",
+                        help="Pass <arg> to the linker",
+                        action="append", default=[])
+    parser.add_argument("-b", dest="backend",
```

Comment:
backend is not sufficiently specified. Think about bmv2, it may support multiple targets. In fact we have 3 factors determining the backend:

1. the architecture (PSA, TSA, etc.)
2. the target (simple switch, FPGA, Tofino, etc.)
3. and the platform (bmv2, cycle accurate model, hardware)

Similar to one of the gnu triples to specify host-os-arch, we need to have an arch-target-platform triple to specify a configuration for which we want to compile, e.g., psa-ss-bmv2.

Also it would be very useful to have an option that list all the supported configurations (the equivalent of --target-help):
```
gcc --target-help
The following options are target specific:
  -m128bit-long-double        sizeof(long double) is 16
  -m16                        Generate 16bit i386 code
  -m32                        Generate 32bit i386 code
  -m3dnow                     Support 3DNow! built-in functions
  -m64                        Generate 64bit x86-64 code
  -m80387                     Use hardware fp
[...]
```